app/models/user.rb

```diff
@@ -353,6 +353,10 @@ def group_granted_trust_level
       .maximum("groups.grant_trust_level")
   end
 
+  def visible_groups
+    groups.select { |group| guardian.can_see_group?(group) }
```

Comment:
```suggestion
    groups.visible_groups(self)
```

I think this works, can you confirm... if passes all tests feel free to merge